dask_planner/src/expression.rs

```diff
@@ -557,214 +557,156 @@ impl PyExpr {
         }
     }
 
-    /// TODO: I can't express how much I dislike explicity listing all of these methods out
-    /// but PyO3 makes it necessary since its annotations cannot be used in trait impl blocks
     #[pyo3(name = "getFloat32Value")]
-    pub fn float_32_value(&mut self) -> PyResult<Option<f32>> {
-        match &self.expr {
-            Expr::Literal(scalar_value) => match scalar_value {
-                ScalarValue::Float32(iv) => Ok(*iv),
-                _ => Err(py_type_err("getValue<T>() - Unexpected value")),
-            },
-            _ => Err(py_type_err("getValue<T>() - Non literal value encountered")),
+    pub fn float_32_value(&self) -> PyResult<Option<f32>> {
+        match self.get_scalar_value()? {
+            ScalarValue::Float32(value) => Ok(*value),
+            other => Err(unexpected_literal_value(other)),
         }
```

Comment:
Is it possible/worthwhile to make a helper function (parametrized on the specific `ScalarValue` variant) for this basic match pattern?

Reply:
That would be easy to do with a macro, but there are pros and cons to using macros, so I planned on doing that in a follow-on PR.

Reply:
We could define this macro:

```rust
macro_rules! extract_scalar_value {
    ($self: expr, $variant: ident) => {
        match $self.get_scalar_value()? {
            ScalarValue::$variant(value) => Ok(*value),
            other => Err(unexpected_literal_value(other)),
        }
    }
}
```

and then most of the get_value methods become one-liners:

```rust
    #[pyo3(name = "getFloat32Value")]
    pub fn float_32_value(&self) -> PyResult<Option<f32>> {
        extract_scalar_value!(self, Float32)
    }
```

The main downside is that there is now another language for the maintainers to become familiar with.

Reply:
Thanks for the detailed response, in that case I'm okay with pushing this off to a follow-up PR, and we can have a large discussion around the use of macros.